compiler/semexprs.nim

```diff
@@ -1754,16 +1754,18 @@ proc semReturn(c: PContext, n: PNode): PNode =
   if c.p.owner.kind in {skConverter, skMethod, skProc, skFunc, skMacro} or
       (not c.p.owner.typ.isNil and isClosureIterator(c.p.owner.typ)):
     if n[0].kind != nkEmpty:
-      # transform ``return expr`` to ``result = expr; return``
       if c.p.resultSym != nil:
-        var a = newNodeI(nkAsgn, n[0].info)
-        a.add newSymNode(c.p.resultSym)
-        a.add n[0]
-        n[0] = semAsgn(c, a)
-        # optimize away ``result = result``:
-        if n[0][1].kind == nkSym and n[0][1].sym == c.p.resultSym:
-          n[0] = c.graph.emptyNode
-      else:
+        # transform ``return expr`` to ``result = expr; return``
+        var asgn = newNodeI(nkAsgn, n[0].info)
+        asgn.add newSymNode(c.p.resultSym)
+        asgn.add n[0]        
+        asgn = semAsgn(c, asgn)
```

Comment:
```suggestion
        let asgn = semAsgn(c, newNodeI(nkAsgn, n[0].info, newSymNode(c.p.resultSym), n[0]))
```
Btw, we now have `newTreeI`/`newTreeIT` similar to `newNodeI` and `newNodeIT` for convinience: https://github.com/nim-lang/Nim/blob/devel/compiler/ast.nim#L1151